### DISCUSSION TITLE
Welcome to rpi-metro-display Discussions!

### DISCUSSION THREAD

**Discussion by kenschneider18** [kenschneider18/rpi-metro-display#3] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by ipk215** [comment]
love this project! I'm definitely a beginner with all of this so I have a couple questions. I have gotten all the way to step 10 where I replace the placeholders. I was wondering if my paths should be brackets, separated by line breaks, or just spaces. I was also wondering for when I'm finished with that do I exit and save the run.sh before moving on to the next step. I originally used spaces, saved, and went on but when I ran the start display nothing happened (no errors, just nothing showing up on the display). Thanks again

**Reply by kenschneider18** [reply]
All you should need are spaces. No brackets like I have around the placeholders `<>` either. You have to save and exit before moving on to step 11.

In regard to running the display and nothing happening, that was a bug discovered by @wellmans. I forgot to add execute permissions to `run.sh`. If you pull the most recent changes from the repo you'll hopefully be good.

**Reply by kenschneider18** [reply]
You might want to try running `sudo ./run.sh` after you pull just to see if the display comes on or gives you any sort of error message.

Post back here if you're still having issues.

**Reply by ipk215** [reply]
Thanks for the quick follow up! Glad to know I was doing step 11 correct. How do I pull the most recent changes?

**Reply by kenschneider18** [reply]
From the directory you cloned `rpi-metro-display`, you'll have to run `git pull`.

Since you've modified `run.sh` you'll probably get a message like the following:

```sh
error: Your local changes to the following files would be overwritten by merge:
        run.sh
Please commit your changes or stash them before you merge.
```

Git is attempting to resolve the changes you made with the new code it's trying to download and can't automatically so you need to get rid of your changes before Git will pull the new code. Obviously we don't want to do that so the way to get around it is to run `git stash`, which is going temporarily hide the changes you've made. Then run `git pull` which will now succeed. Then to get your changes back run `git stash pop`.

That should hopefully be it.

**Reply by ipk215** [reply]
Ok I ran git pull from the rpi-metro-display directory:

root@DietPi:~/metro-sign/rpi-metro-display# git pull
Already up to date.

I didn't get any errors but I'm not getting any signs of life from the led screens after I ran the start line. Before I started this project I tested them out so I know they do work and are wired correctly.

**Reply by kenschneider18** [reply]
What's the output of running `git status`?

**Reply by ipk215** [reply]
root@DietPi:~/metro-sign/rpi-metro-display# git status
On branch master
Your branch is up to date with 'origin/master'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

        modified:   run.sh

no changes added to commit (use "git add" and/or "git commit -a")
root@DietPi:~/metro-sign/rpi-metro-display#

**Reply by kenschneider18** [reply]
I doubt it's an issue with your display. I'm almost positive it's a permissions issue. You can test that though. If you copy the python command you modified from `run.sh` and run it as root a.k.a.

```sh
sudo python <the rest of your command here>
```

The display should turn on. Really all that `run.sh` script does is run this command for you without making you type out that whole python command every time.

**Reply by kenschneider18** [reply]
Interesting, when you run `git log` does the first section show this commit?

```
commit 064354f798e6add8e99ee701854547dc304e69f4 (HEAD -> master, tag: v1.0.3, origin/master, origin/HEAD)
Author: Ken Schneider <ken.schneider18@gmail.com>
Date:   Sat Mar 13 20:39:29 2021 -0500

    Add execute permissions to run.sh (#10)
```

**Reply by wellmans** [reply]
Set to execute permission on run.sh and give it a try. Mine is booting directly to the display now that was resolved.

**Reply by ipk215** [reply]
When I run git log I do get that as the first section. I ran the command as root "sudo python /home/dietpi...." I got this error:

python: can't open file '/home/dietpi/metro-sign/rpi-metro-display/rpi-metro-display.py': [Errno 2] No such file or directory

I don't get this error (or any) when using "sudo systemctl start metro-display"

**Reply by kenschneider18** [reply]
Yeah, sounds like one of my TODOs is going to be adding a step where you test the set up to make sure your paths are all good. What that message means is the path you have there is incorrect for what's on your system. What's the output of `pwd` from the code directory?

**Reply by ipk215** [reply]
Hmm ok not sure where I would have messed up the path. When I use pwd I get:

root@DietPi:~/metro-sign/rpi-metro-display# pwd
/root/metro-sign/rpi-metro-display

**Reply by kenschneider18** [reply]
It's easy enough to do. It doesn't really make a difference except that you need to be aware of it when you configure `run.sh` or try to run it with the canonical path (which is what we need to do in `run.sh`) rather than relative. 

So `pwd` gives you the canonical path to your current directory. So since in your case the location of the folder containing `rpi-metro-display.py` is `/root/metro-sign/rpi-metro-display/` you'll need to change that.

So your python command becomes:

```sh
sudo python /root/metro-sign/rpi-metro-display/rpi-metro-display.py
```

Same thing applies for any of the other files that you've used the "default" path for `/home/dietpi/metro-sign/<rest-of-path>` in the other placeholders.

Give that a whirl just running the python command outside `run.sh` before you try `run.sh`. That's correct, the error message won't show up when you're using `systemctl`.

**Reply by ipk215** [reply]
Ok I replaced the home/dietpi with root for all my placeholders, ran the program outside of run.sh and got this error.


root@DietPi:~/metro-sign/rpi-metro-display# sudo python /root/metro-sign/rpi-metro-display/rpi-metro-display.py /root/metro-sign/rpi-metro-display/log.txt eb044b16f6c147098f93d954698944fb K04 1 /root/metro-sign/rpi-metro-display/6x10.bdf /root/metro-sign/rpi-metro-display/lines.json /root/metro-sign/rpi-metro-display/stations.json
 * Serving Flask app "rpi-metro-display" (lazy loading)
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off
Process Process-2:
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/process.py", line 267, in _bootstrap
    self.run()
  File "/usr/lib/python2.7/multiprocessing/process.py", line 114, in run
    self._target(*self._args, **self._kwargs)
  File "/root/metro-sign/rpi-metro-display/rpi-metro-display.py", line 82, in run_display
  File "/root/metro-sign/rpi-metro-display/rpi-metro-display.py", line 192, in draw_display
  File "graphics.pyx", line 32, in rgbmatrix.graphics.Font.LoadFont
Exception: Couldn't load font /root/metro-sign/rpi-metro-display/6x10.bdf

**Reply by kenschneider18** [reply]
`Exception: Couldn't load font /root/metro-sign/rpi-metro-display/6x10.bdf` Is that font file in that location? If you `ls` from `rpi-metro-display` is `6x10.bdf` there?

**Reply by kenschneider18** [reply]
If not you might need to re-do step 9

**Reply by ipk215** [reply]
> Set to execute permission on run.sh and give it a try. Mine is booting directly to the display now that was resolved.

giving this another try, I'm not entirely sure how to set to execute permission on run.sh.

**Reply by kenschneider18** [reply]
If you pull the latest code it should be set, regardless you can force it by running:
```sh
chmod u+x (path to run.sh here)
```

**Comment by namtendo** [comment]
I'm also stuck at step 10. For the "station code" and "direction code" parts I'm not sure what to put. Is there a standard list of station codes to use? And do I put the value in quotation marks? I'm trying to get it to show the navy yard station in both directions but putting "Navy Yard" for station code doesn't seem to work.

**Reply by namtendo** [reply]
Alright, I got it to work by using the station codes from the JSON file. My next question is, is it possible to make it display times for both directions? Currently I'm only able to make it show the Greenbelt direction.

**Reply by kenschneider18** [reply]
The direction code that you pass in your `run.sh` defines which direction it’s going in. It’s either 1 or 2. So you can switch the default that way. Once it’s running though, you can hit the [API from another computer to change it without even using SSH.](https://github.com/kenschneider18/rpi-metro-display#put-stationname). You can change the direction and station this way whenever you want.

Now, if you want to see trains for each direction on that same screen at the same time? That’s doable, but you’d have to modify the code because I wrote it to display one side of the tracks at a time. It’s not a terrible idea to make that an option though. It’d actually be a [one line change](https://github.com/kenschneider18/rpi-metro-display/blob/a6205d0301f6f02e6355210f1c36e23f3a773d70/rpi-metro-display.py#L137). If this is what you’re interested in let me know.

**Reply by namtendo** [reply]
Oh yes I would be interested in having that option to show both directions. I'll try looking at the code but my programming background isn't the strongest so I'm a little bit weary of making any changes myself.

Also, as a note for anyone trying to hit the API using those commands from Windows, it won't working using the command prompt because of quirks specific to WIndows, (e.g. it really doesn't like single quotes). I got it to work by downloading and running it with Cygwin.

**Reply by namtendo** [reply]
Actually, I assume I just have to comment out that line and get rid of the if condition?

**Reply by kenschneider18** [reply]
You could try that. The thing that I worry about is the display can only show 3 trains at a time so I’m not sure what will happen when you start trying to display the 4th train (the API sends back a maximum of 6, 3 per direction/group). To be safe I would add a counter or something to ensure that you only return the first 3 trains.

**Reply by namtendo** [reply]
Thanks for the help! It runs well and looks really cool on top of the bookshelf in my living room. I first removed the if condition and it seems to run just fine showing 3 trains. I did add a condition to limit it to 3 later on just to be safe and it works the same. Was it difficult making the wooden frame for the display? I may do that as the next step in this project when I have time.

**Reply by kenschneider18** [reply]
Glad to hear it. Yeah essentially what it’s going to do is draw those trains further down on the display, but since the coordinates don’t exist in the dot matrix I guess nothing happens. Still not a bad idea to narrow it down to 3 just to be safe.

Not as difficult as I thought it would be, but I had help. The main frame was just off the shelf barn style wood from Home Depot. I took that over to Ace and they were able to help turn that into a plan and cut everything. Even so, the cuts were off so I basically had to glue the frame together with the displays inside. It’s not super pretty up close but it does the job.
